### PR TITLE
api update for sm3 and md5

### PR DESCRIPTION
**Pull request by notch1p**
updated api to match that of rust. see the updated `crypto/README.md` for usage. Also switch to iter[byte] for SM3 internally, which should perform better.

### CODE REVIEW

File: crypto/README.md [hunk 1]
```diff
@@ -12,12 +12,39 @@ A collection of cryptographic hash functions and utilities.
...
+let input = "The quick brown fox jumps over the lazy dog"
+println(bytes_to_hex_string(sm3sum(input.to_bytes())))
+// => fc2b31896629e88652ca1e3be449ec7ec93f7e5e29769f273fb973bc1858c66d
+
+
+//buffered
+let ctx = SM3Context::make()
+ctx.update(b"\x61") // 'a'
+ctx.update(b"\x62") // 'b'
+ctx.update(b"\x63") // 'c'
```
Comment (peter-jerry-ye) [inline crypto/README.md:47]:
```suggestion
ctx.update(b"a")
ctx.update(b"b")
ctx.update(b"c")
```

File: crypto/README.md [hunk 2]
```diff
@@ -12,12 +12,39 @@ A collection of cryptographic hash functions and utilities.
...
+### SM3
+
+```moonbit
+let input = "The quick brown fox jumps over the lazy dog"
+println(bytes_to_hex_string(sm3sum(input.to_bytes())))
+// => fc2b31896629e88652ca1e3be449ec7ec93f7e5e29769f273fb973bc1858c66d
+
+
+//buffered
+let ctx = SM3Context::make()
```
Comment (peter-jerry-ye) [inline crypto/README.md:44]:
```suggestion
let ctx = SM3Context::new()
```

File: crypto/crypto.mbti [hunk 3]
```diff
@@ -1,25 +1,31 @@
...
 fn chacha12(FixedArray[UInt], UInt, Bytes, ~nonce : UInt = ..) -> Bytes!
 
 fn chacha20(FixedArray[UInt], UInt, Bytes, ~nonce : UInt = ..) -> Bytes!
 
 fn chacha8(FixedArray[UInt], UInt, Bytes, ~nonce : UInt = ..) -> Bytes!
 
-fn md5(Bytes) -> Bytes
+fn md5sum(Bytes) -> Bytes
+
+fn rotate_left(Int, Int) -> Int
```
Comment (peter-jerry-ye) [inline crypto/crypto.mbti:20]: Why is it public?

File: crypto/crypto.mbti [hunk 4]
```diff
@@ -1,25 +1,31 @@
...
 fn sha1(Bytes) -> Bytes
 
-fn sm3(Bytes) -> Array[UInt]
+fn sm3sum(Bytes) -> Bytes
+
+fn sm3sum_from_iter(Iter[Byte]) -> Bytes
+
+fn u32_to_u8be(UInt) -> Array[Byte]
 
 fn u8_to_u32be(Bytes, ~i : Int = ..) -> UInt
```
Comment (peter-jerry-ye) [inline crypto/crypto.mbti:30]: Why are they public?

File: crypto/crypto.mbti [hunk 5]
```diff
@@ -28,6 +34,22 @@ fn u8_to_u32le(Bytes, ~i : Int = ..) -> UInt
 fn uints_to_hex_string(Array[UInt]) -> String
 
 // Types and methods
+type MD5Context
+impl MD5Context {
+  compute(Self) -> Bytes
+  finalize(Self) -> Bytes
+  make() -> Self
```
Comment (peter-jerry-ye) [inline crypto/crypto.mbti:41]:
```suggestion
  new() -> Self
```

File: crypto/crypto.mbti [hunk 6]
```diff
@@ -28,6 +34,22 @@ fn u8_to_u32le(Bytes, ~i : Int = ..) -> UInt
...
+  update(Self, Bytes) -> Unit
+}
+
+type SM3Context
+impl SM3Context {
+  compute(Self) -> Bytes
+  finalize(Self) -> Bytes
+  make() -> Self
```
Comment (peter-jerry-ye) [inline crypto/crypto.mbti:49]:
```suggestion
  new() -> Self
```

File: crypto/md5.mbt [hunk 7]
```diff
@@ -219,16 +231,28 @@ fn md5_transform(state : FixedArray[UInt], input : FixedArray[UInt]) -> Unit {
...
     bytes_to_hex_string(hash),
     content="b0986ae6ee1eefee8a4a399090126837",
   )
 }
+
+test {
+  let ctx = MD5Context::make()
+  md5_update(ctx, b"\x61")
+  md5_update(ctx, b"\x62")
+  md5_update(ctx, b"\x63")
```
Comment (peter-jerry-ye) [inline crypto/md5.mbt:252]: Same as above

File: crypto/crypto.mbti [hunk 8]
```diff
@@ -1,25 +1,31 @@
...
+fn bytes_to_iter(Bytes) -> Iter[Byte]
+
 fn chacha12(FixedArray[UInt], UInt, Bytes, ~nonce : UInt = ..) -> Bytes!
 
 fn chacha20(FixedArray[UInt], UInt, Bytes, ~nonce : UInt = ..) -> Bytes!
 
 fn chacha8(FixedArray[UInt], UInt, Bytes, ~nonce : UInt = ..) -> Bytes!
 
-fn md5(Bytes) -> Bytes
+fn md5sum(Bytes) -> Bytes
```
Comment (peter-jerry-ye) [inline crypto/crypto.mbti:18]: No need for renaming

File: crypto/sm3.mbt [hunk 9]
```diff
@@ -16,20 +16,21 @@
 // - [GM/T 0004-2012]    https://www.oscca.gov.cn/sca/xxgk/2010-12/17/1002389/files/302a3ada057c4a73830536d03e683110.pdf
 // SM3 is as secure as SHA256, providing similar performance. https://doi.org/10.3390/electronics8091033
 
-priv struct SM3Context {
+struct SM3Context {
   reg : FixedArray[UInt] // register A B C D E F G H. i.e. digest
   mut len : UInt64
-  mut msg : Array[Byte]
+  mut msg : Iter[Byte]
```
Comment (peter-jerry-ye) [inline crypto/sm3.mbt:22]: Why save an `Iter`?

File: crypto/sm3.mbt [hunk 10]
```diff
@@ -77,23 +78,23 @@ fn SM3Context::p_1(x : UInt) -> UInt {
   x ^ urotate_left(x, 15) ^ urotate_left(x, 23)
 }
 
-fn pad(self : SM3Context) -> Array[Byte] {
-  self.msg.push(b'\x80')
-  while self.msg.length() % 64 != 56 {
-    self.msg.push(b'\x00')
+fn pad(self : SM3Context) -> Iter[Byte] {
+  self.msg = self.msg.append(b'\x80')
+  while self.msg.count() % 64 != 56 {
```
Comment (peter-jerry-ye) [inline crypto/sm3.mbt:83]: This is inefficient: the iter will execute everytime when you count

File: crypto/crypto.mbti (hunk 3 above)
Comment (notch1p) [reply crypto/crypto.mbti:20]: I think rotate_left could be useful in some other scenarios, thus making it public

File: crypto/crypto.mbti (hunk 4 above)
Comment (notch1p) [reply crypto/crypto.mbti:30]: same as above

File: crypto/sm3.mbt (hunk 10 above)
Comment (notch1p) [reply crypto/sm3.mbt:83]: changed to this
```moonbit
  let mut cnt = self.msg.count()
  while cnt % 64 != 56 {
    self.msg = self.msg.append(b'\x00')
    cnt += 1
  }
```

File: crypto/sm3.mbt (hunk 9 above)
Comment (notch1p) [reply crypto/sm3.mbt:22]: currently `update()` `update_from_iter()` accepts `data` of type `Bytes` or `Iter[Byte]`. `data` is used to populate `msg`. and is further used in `sm3_transform`. Saving an Iter[Byte] means Bytes can be convert to Iter[Byte] at no cost, but not Array[Byte]

**Pull request merged**